apicast/src/plugin.lua

```diff
@@ -0,0 +1,57 @@
+-- Apicast provides a plugin system. To add a custom plugin, we only need to
+-- define a module that implements a few methods like rewrite(), access(), etc.
+-- that are called in the appropriate nginx phases.
+-- The goal of this lua module is to implement the methods that are not likely
+-- to be changed from plugin to plugin so we do not have to repeated code
+-- across all of them.
+
+local balancer = require 'balancer'
+
+-- Plugins need access to provider and configuration, so we need to expose them
+local _M = { provider = require 'provider',
```

Comment:
I don't know. I think both provider and configuration are badly defined and scoped.
Exposing them makes me feel dirty. The way provider stores configuration is specific to that and I'd not like much to expose it.